R/as_model_out_df.R

```diff
@@ -0,0 +1,211 @@
+#' @param sep character string. Character used as separator when concatenating
+#' `team_abbr` and `model_abbr` column values into a single `model_id` string. Only
+#' applicable if `model_id` column not present and `team_abbr` and `model_abbr` columns are.
+#' @param trim_to_task_ids logical. Whether to trim `tbl` to task ID columns only. Task ID
+#' columns can be specified by providing a `<hub_connection>` class object to `hub_con` or
+#' manually through `task_id_cols`.
+#' @param hub_con a `<hub_connection>` class object. Only used if
+#' `trim_to_task_ids = TRUE` and tasks IDs should be determined from the hub config.
+#' @param task_id_cols a character vector of column names. Only used if
+#' `trim_to_task_ids = TRUE` to manually specify task ID columns to retain.
+#' Overrides `hub_con` argument if provided.
+#' @param remove_empty Logical. Whether to remove columns containing only `NA`.
+#'
+#' @return A `model_out_tbl` class object.
+#' @export
+#'
+#' @examples
+#' library(dplyr)
+#' hub_path <- system.file("testhubs/flusight", package = "hubUtils")
+#' hub_con <- connect_hub(hub_path)
+#' hub_con %>%
+#'   filter(output_type == "quantile", location == "US") %>%
+#'   collect() %>%
+#'   filter(forecast_date == max(forecast_date)) %>%
+#'   as_model_out_tbl()
+#'
+as_model_out_tbl <- function(tbl, model_id_col = NULL, output_type_col = NULL,
+                             output_type_id_col = NULL, value_col = NULL, sep = "-",
+                             trim_to_task_ids = FALSE, hub_con = NULL,
+                             task_id_cols = NULL, remove_empty = FALSE) {
+  checkmate::assert_data_frame(tbl)
+  tbl <- tibble::as_tibble(tbl)
+
+  tbl <- rename_columns(tbl,
+    model_id_col = model_id_col,
+    output_type_col = output_type_col,
+    output_type_id_col = output_type_id_col,
+    value_col = value_col
+  )
+
+  if (!"model_id" %in% names(tbl)) {
+    cli::cli_alert_warning("{.arg model_id} column missing. Attempting to create automatically.")
+    tbl <- model_id_merge(tbl, sep = sep)
+  }
+
+  if (trim_to_task_ids) {
+    tbl <- trim_tbl_to_task_ids(
+      tbl = tbl,
+      task_id_cols = task_id_cols,
+      hub_con = hub_con
+    )
+  }
+
+  if (remove_empty) {
+    tbl <- remove_model_out_tbl_empty(tbl)
+  }
+
+  tbl <- std_col_order_model_out_tbl(tbl) %>%
+    std_row_order_model_out_tbl()
+
+  validate_model_out_tbl(tbl)
+
+  structure(tbl,
+    class = c("model_out_tbl", class(tbl))
+  )
+}
+
+# --- Unexported Utilities ----
+validate_model_out_tbl <- function(tbl) {
+  if (!all(std_colnames %in% names(tbl))) {
+    cli::cli_abort(
+      c(
+        "x" = "Standard column{?s} {.val
+              {std_colnames[!std_colnames %in% names(tbl)]}}
+              missing from {.arg tbl}.",
+        "!" = "Valid {.cls model_out_tbl} objects must contain standard
+              columns {.val {std_colnames}}"
+      )
+    )
+  }
+  if (nrow(tbl) == 0) {
+    cli::cli_warn(c("!" = "{.arg tbl} has zero rows."))
+  }
```

Comment:
I'm feeling hesitant about whether we want to throw a warning here? On one hand, seems reasonable to throw a warning as most of the time ending up with a zero-row data frame of model outputs will be an accident. On the other hand, most data.frame-like objects let you create a frame with 0 rows without errors/warnings, and there may be reasonable workflows where this comes up (e.g. maybe somehow loading forecasts by model and location for processing?)

Reply:
Totally up to you. I agree that a zero row tibble is totally reasonable. Having said that we are converting to the class because we want to standardise and ensure validity of objects for downstream hubverse processes which, my impression is that, in the vast majority of cases will require data and fail if given zero rows?

It is a warning in any case and will not stop the flow of operations, just issue formal notification of the issue. 

One thing we could do is downgrade to a message is you like (i.e. use `cli_alert_warning()` instead of `cli_warn()`?

Otherwise it is indeed also a check that could just be included in the downstream process functions (e.g. as part of `simple_ensemble()`) that require data with > 0 rows.

Reply:
- I'd vote for either issuing a warning, as you have now, or no output, not a message. Either we think this is a problem or we don't; generating an informational message doesn't seem worthwhile.
- Still no strong preference about this, so let's just leave it as is unless we decide otherwise some time later on.